### ISSUE TITLE
FirebaseMessagingException is retried but not recovered

### ISSUE THREAD

**Issue by mbdeveci** [spring-projects/spring-retry#233]
Spring Boot: 2.3.6
Kotlin: 1.3.72 
Spring-retry: 1.2.5

I have been trying to send notifications with firebase and I wanted to use retry-recover 

```kotlin
@Retryable(value = [FirebaseMessagingException::class], backoff = Backoff(delay = 100))
fun sendNotification() { // impl.}

@Recover
    fun recover(ex: FirebaseMessagingException) { //recover}
```
when i do this, i can see that its retried 3 times on failure and i get exception with message: Cannot locate recovery method

but when i do 
```kotlin
@Retryable(value = [DummyException::class], backoff = Backoff(delay = 100))
fun sendNotification() { 
    try {
       // function call that throws FirebaseMessagingException
    }
    catch(ex: FirebaseMessagingException) { 
           throw DummyException()
    }
}

@Recover
fun recover(ex: DummyException) { //recover}
```
it retries 3 times and recover function is triggered. I wonder what can be the cause ?

**Comment by garyrussell** [comment]
GitHub issues are not for asking questions; use Stack Overflow or [Discussions](https://github.com/spring-projects/spring-retry/discussions) for questions.

Issues are for reporting bugs or asking for features.

I am surprised the retries work - CGLib can't proxy final methods - they didn't for me.

This works fine for me (with spring-retry 1.3.1)...

```kotlin
@Component
open class Foo {

    @Retryable(IllegalStateException::class)
    open fun canRetry() {
        println("here")
        throw IllegalStateException("test")
    }

    @Recover
    open fun recovered(ex: IllegalStateException) {
        println("recovered: " + ex.message)
    }

}
```
```
here
here
here
recovered: test
```

**Comment by garyrussell** [comment]
If you are using an interface (which is implied by the fact you say the retries work), versions before 1.3.0.RELEASE did not find `@Recover` methods on the target class.

https://github.com/spring-projects/spring-retry/commit/8f3ceedd29d42b04c157e9fc21d4f5426cd71cdb

**Comment by mbdeveci** [comment]
sorry for the issue :/ Yes im using an interface but same methods work with dummy exception.
I upgraded my version to 1.3.1 and the situation is not changed. I could not understand which part is incorrect, when I change handled exception to something else retry and recover works with 1.2.5, when i make dummy exception final class it still works.  (FirebaseMessagingException is final class)
I solved my problem by catching and throwing another exception, maybe problem is related with firebase library. Anyway if you want to dig deeper let me know. Thanks for helping
